Basic comments handling

ludwigjossieaux opened the pull request:
Hello,

Related to issue #271, please find a basic handling of the comments at step level. 

Based on the location (line, column) provided by the parser, the comments above the steps are considered commenting the step below. There is some edge cases not yet taken into account (comments below the last step, for instance); if you agree with the basic principle, I can make sure these are also taken into account. 

There are also few tests checking the feature is working for the output formats.

dirkrombauts commented:
Cool, I like it. I had a quick review of the code and I think the overall direction is good. So please go ahead and add handling for the edge cases you mentioned.

One concern - it's a bit of a mystery to me: the continuous build successfully built your pull request, but when I checked it out locally and ran the tests, two of them were failing. Both of them in the FormattingAFeature file. The reason is that the expected JSON did not include arrays for the comments. Could you check that in your version please!

And again, thanks for this big contribution. It would be cool if you could finalize it this week so it can be released in time for the CukeUP conference next week :-)

ludwigjossieaux commented:
Thanks for the feedback ! it's doable this week, I'll add the missing commits.

Concerning the two tests, you are right it is missing the comment arrays on the step object, it will be added with the next commits. However these tests were already failing with the fork I did (still the head revision on develop). It looks like the "Summary" object is missing between "Features" and "Configuration". Is it correct ?

dirkrombauts commented:
I fixed the failing tests (by removing the "configuration" line which was an overspecification of the test). It's a bit of a mystery why the test sometimes passes and sometimes fails. I have my suspicions but no proof :-/

Anyway, you can now pull the latest commit from picklesdoc/develop with the fixed test.

ludwigjossieaux commented:
Hi,
I've pushed the commits for the comments located after the last step of a scenario. I'm still not entirely sure how handle more freely placed comments (outside of steps). I've also added some more tests and fixed those not passing.
Let me know if would think of something else I could add

dirkrombauts commented on src/Pickles/Pickles.ObjectModel/ObjectModel/Comment.cs:
```diff
@@ -0,0 +1,48 @@
+            // Set default
+            this.Type = CommentType.Normal;
+        }
+        
```
This line contains whitespace, please remove it

dirkrombauts commented on src/Pickles/Pickles.ObjectModel/ObjectModel/Comment.cs:
```diff
@@ -0,0 +1,48 @@
+//  </copyright>
+//  --------------------------------------------------------------------------------------------------------------------
+
+using System.Collections.Generic;
```
please remove the unused using

dirkrombauts commented on src/Pickles/Pickles.ObjectModel/ObjectModel/Location.cs:
```diff
@@ -0,0 +1,31 @@
+//  </copyright>
+//  --------------------------------------------------------------------------------------------------------------------
+
+using System.Collections.Generic;
```
Please remove the unused using.

dirkrombauts commented on src/Pickles/Pickles.ObjectModel/ObjectModel/Scenario.cs:
```diff
@@ -44,6 +44,8 @@ public Scenario()
 
+        public Location Location { get; set; }
+
         #endregion
```
This is not strictly your code, but do me a favour and remove the `#region` from the class. Thanks!

dirkrombauts commented on src/Pickles/Pickles.ObjectModel/ObjectModel/Step.cs:
```diff
@@ -31,5 +38,9 @@ public class Step
+
+        public Location Location { get; set; }
+
+        public List<Comment> Comments { get; set; } 
```
Please remove the extra whitespace at the end of the `Comments { get; set; }` line

dirkrombauts commented on src/Pickles/Pickles.Test/DocumentationBuilders/Excel/WhenAddingAStepToAWorksheet.cs:
```diff
@@ -47,5 +49,107 @@ public void ThenStepAddedSuccessfully()
         }
+
+        [Test]
+        public void ThenStepCommentsAreAddedSuccesfully()
```
Successfully is spelled with three s in total :-)

dirkrombauts commented on src/Pickles/Pickles.Test/DocumentationBuilders/Excel/WhenAddingAStepToAWorksheet.cs:
```diff
@@ -47,5 +49,107 @@ public void ThenStepAddedSuccessfully()
+        }
+
+        [Test]
+        public void ThenMultilineStepCommentsAreAddedSuccesfully()
```
Successfully is spelled with three s in total :-)

dirkrombauts commented on src/Pickles/Pickles.Test/DocumentationBuilders/Excel/WhenAddingAStepToAWorksheet.cs:
```diff
@@ -47,5 +49,107 @@ public void ThenStepAddedSuccessfully()
+        }
+
+        [Test]
+        public void ThenCommentsAfterTheLastStepAreAddedSuccesfully()
```
Successfully is spelled with three s in total :-)

dirkrombauts commented on src/Pickles/Pickles.Test/DocumentationBuilders/HTML/WhenFormattingStep.cs:
```diff
@@ -181,5 +181,123 @@ public void Tables_are_formatted_as_list_items_with_tables_internal()
+                Name = "a simple step",
+                TableArgument = null,
+                DocStringArgument = null,
+                Comments = new List<Comment>()
```
the `()` is unnecessary.

dirkrombauts commented on src/Pickles/Pickles.Test/DocumentationBuilders/HTML/WhenFormattingStep.cs:
```diff
@@ -181,5 +181,123 @@ public void Tables_are_formatted_as_list_items_with_tables_internal()
+                DocStringArgument = null,
+                Comments = new List<Comment>()
+                {
+                    new Comment()
```
the `()` is unnecessary.

dirkrombauts commented on src/Pickles/Pickles.Test/DocumentationBuilders/HTML/WhenFormattingStep.cs:
```diff
@@ -181,5 +181,123 @@ public void Tables_are_formatted_as_list_items_with_tables_internal()
+            var formatter = Container.Resolve<HtmlStepFormatter>();
+            XElement actual = formatter.Format(step);
+
+            XNamespace xmlns = XNamespace.Get("http://www.w3.org/1999/xhtml");
```
Do me a favor and introduce a private field with the namespace and use it in all tests in this class. Thanks!

dirkrombauts commented on src/Pickles/Pickles.Test/DocumentationBuilders/HTML/WhenFormattingStep.cs:
```diff
@@ -181,5 +181,123 @@ public void Tables_are_formatted_as_list_items_with_tables_internal()
+                Name = "a simple step",
+                TableArgument = null,
+                DocStringArgument = null,
+                Comments = new List<Comment>()
```
the `()` is unnecessary.

dirkrombauts commented on src/Pickles/Pickles.Test/DocumentationBuilders/HTML/WhenFormattingStep.cs:
```diff
@@ -181,5 +181,123 @@ public void Tables_are_formatted_as_list_items_with_tables_internal()
+                DocStringArgument = null,
+                Comments = new List<Comment>()
+                {
+                    new Comment()
```
the `()` is unnecessary.

dirkrombauts commented on src/Pickles/Pickles.Test/DocumentationBuilders/HTML/WhenFormattingStep.cs:
```diff
@@ -181,5 +181,123 @@ public void Tables_are_formatted_as_list_items_with_tables_internal()
+                        Text = "    # A simple comment",
+                        Type = CommentType.StepComment
+                    },
+                    new Comment()
```
the `()` is unnecessary.

dirkrombauts commented on src/Pickles/Pickles.Test/DocumentationBuilders/HTML/WhenFormattingStep.cs:
```diff
@@ -181,5 +181,123 @@ public void Tables_are_formatted_as_list_items_with_tables_internal()
+                Name = "a simple step",
+                TableArgument = null,
+                DocStringArgument = null,
+                Comments = new List<Comment>()
```
the `()` is unnecessary.

dirkrombauts replied:
the `()` is unnecessary.

dirkrombauts commented on src/Pickles/Pickles.Test/DocumentationBuilders/HTML/WhenFormattingStep.cs:
```diff
@@ -181,5 +181,123 @@ public void Tables_are_formatted_as_list_items_with_tables_internal()
+                DocStringArgument = null,
+                Comments = new List<Comment>()
+                {
+                    new Comment()
```
the `()` is unnecessary.

dirkrombauts commented on src/Pickles/Pickles.Test/DocumentationBuilders/HTML/WhenFormattingStep.cs:
```diff
@@ -181,5 +181,123 @@ public void Tables_are_formatted_as_list_items_with_tables_internal()
+                        Text = "    # A first line",
+                        Type = CommentType.StepComment
+                    },
+                    new Comment()
```
the `()` is unnecessary.

dirkrombauts commented on src/Pickles/Pickles/DocumentationBuilders/HTML/HtmlStepFormatter.cs:
```diff
@@ -19,8 +19,11 @@
+using System.Text;
 using System.Xml.Linq;
-
+using DocumentFormat.OpenXml.Office.Word;
```
Please remove this unused using - especially since Word and Html are different things :-)

dirkrombauts commented on src/Pickles/Pickles/DocumentationBuilders/HTML/HtmlStepFormatter.cs:
```diff
@@ -19,8 +19,11 @@
 
 using System;
+using System.Collections.Generic;
+using System.Linq;
```
Please remove this unused using

dirkrombauts commented on src/Pickles/Pickles/DocumentationBuilders/HTML/HtmlStepFormatter.cs:
```diff
@@ -19,8 +19,11 @@
 //  --------------------------------------------------------------------------------------------------------------------
 
 using System;
+using System.Collections.Generic;
```
Please remove this unused using

dirkrombauts commented on src/Pickles/Pickles/ObjectModel/Mapper.cs:
```diff
@@ -38,7 +38,7 @@ public Mapper(string featureLanguage = LanguageServices.DefaultLanguage)
 
             configurationStore.CreateMap<string, Keyword>().ConvertUsing(new KeywordResolver(featureLanguage));
-
+            
```
Please remove the whitespace in the line

dirkrombauts commented on src/Pickles/Pickles/ObjectModel/Mapper.cs:
```diff
@@ -97,9 +114,42 @@ public Mapper(string featureLanguage = LanguageServices.DefaultLanguage)
+                                if (relatedFeatureElement != null)
+                                {
+                                    var stepAfterComment = relatedFeatureElement.Steps.FirstOrDefault(x => x.Location.Line > comment.Location.Line);
+                                    
```
Please remove the whitespace in the line

dirkrombauts commented:
Thanks for your continuing efforts!

I added a bunch of codestyle-related remarks to the files.

I would also like you to write unit tests for the mapper, like the once you find in https://github.com/picklesdoc/pickles/tree/develop/src/Pickles/Pickles.Test/ObjectModel

ludwigjossieaux commented:
Hi, I've just finished the clean up + the test missing for the mapper, hopefully I didn't miss anything.

dirkrombauts commented:
So far so good - thanks!

I also meant really granular tests where you test the mapping of Comment and Location in isolation, like for example the steps class mapping is being tested in https://github.com/picklesdoc/pickles/blob/develop/src/Pickles/Pickles.Test/ObjectModel/MapperTestsForStep.cs.

ludwigjossieaux commented:
Ah alright, I've misunderstood the request, I've added two tests classes for comment and location, that should cover these cases

dirkrombauts commented:
Thanks for your time and energy!

dirkrombauts commented:
Released in version [2.6.0](https://github.com/picklesdoc/pickles/releases/tag/v2.6.0)
